Song select

ddevault opened the pull request:
![](https://sr.ht/0Xtf.png)

ddevault committed:
Improve song selection layout, database loading

Also adds event that notifes the song select when a beatmap is added.

ddevault committed:
Adjust PlaySongSelect to match UI standards

peppy commented:
I know this is a WIP, but curious how you made this work (Game is almost 100% guaranteed to be null there)

![image](https://cloud.githubusercontent.com/assets/191335/19299894/038f2e5c-9092-11e6-9f7c-ffca1a387776.png)

SirCmpwn commented:
Did you rebase onto this branch? It works fine for me, at least before whatever changes you're working on.

ddevault committed:
Fix up song select based on upstream changes

ddevault committed:
Add test case for song selection

ddevault committed:
Add selection interactions

ddevault committed:
Add background to song select

ddevault committed:
Tweak layout of beatmap groups

ddevault committed:
Get some more of the design's layout in place

ddevault committed:
Use beatmap author in listing

SirCmpwn commented:
![](https://sr.ht/xnxF.png)

ddevault committed:
Fix width of beatmaps

ddevault committed:
Fix up padding on song select

@peppy btw scrolling beyond the ends of the ScrollContainer behaves
weirdly for this container

peppy committed:
Don't bail on no beatmaps in database.

ddevault committed:
Don't crash if you leave and return to song select

During visual tests

ddevault committed:
Refactor beatmap import secondary process

Doesn't launch a new game window and now supports several files at once.

ddevault committed:
Pull beatmap list from db and render simple list

ddevault committed:
Fix up wrt upstream development

ddevault committed:
Fix issues with beatmap import file contention

ddevault committed:
Fix cascade insert

ddevault committed:
Fix initialization of BeatmapInfo

ddevault committed:
Fix additional bugs, tweak L+F

ddevault committed:
Add background textures to beatmap sets

Needs osu-framework#189

ddevault committed:
Add glow to beatmap groups

ddevault committed:
Add beatmap difficulty selection logic

ddevault committed:
Tweak play button

ddevault committed:
Select first difficulty by default

ddevault committed:
Fix assertion failures

ddevault committed:
Remove dependency on DeferredSprite

ddevault committed:
Increase spacing to match mockups

ddevault committed:
Add animation, selection indicator to difficulties

ddevault committed:
Add star counter to difficulty select

ddevault committed:
Sort by difficulty and fix relationship

peppy committed:
Fix cascading inserts.

peppy committed:
Don't insert to database until we're completely finished importing a beatmap.

peppy committed:
Add a database reset method.

peppy committed:
Return the populated item when calling GetChildren to write more elegant code.

peppy committed:
Make importing work properly.

Moves import code to BeatmapDatabase.

SirCmpwn commented on osu.Game/Database/BeatmapDatabase.cs:
```diff
@@ -139,9 +139,12 @@ public Beatmap GetBeatmap(BeatmapInfo beatmapInfo)
+            if (item == null) return default(T);
+
+            connection.GetChildren(item, recursive);
+            return item;
```
I feel like this change is misleading. It implies that you're getting a new object back and that this is a pure function, neither of which is the case.

ddevault committed:
Implement Play button, drop textures

ddevault committed:
Implement handoff to Player

ddevault committed:
Implement fixes related to auto size changes

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,249 @@
+                {
+                    RelativeSizeAxes = Axes.X,
+                    AutoSizeAxes = Axes.Y,
+                    Width = 1,
```
unnecessary (1 is default)

SirCmpwn replied:
ty

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,249 @@
+                    {
+                        setBox = new BeatmapSetBox(beatmapSet)
+                        {
+                            RelativeSizeAxes = Axes.X,
```
i'd probably move all of these inside the class's ctor.

SirCmpwn replied:
I wouldn't. I try to make the judgement call about which of these properties is specific to *this* place the component is being used, and which is specific to the component itself. If we were to render a BeatmapBox elsewhere, we would likely want to use a different layout.

ddevault committed:
Fix issues with invisible beatmaps

ddevault committed:
Fix sizing of setList flow container

ddevault committed:
Fix auto size on BeatmapSetBox difficulties

ddevault committed:
Fix busted test project

ddevault committed:
Add test data to db for VisualTests

Also fixes the broken IPC condition

ddevault committed:
Fix up JSON reference

ddevault committed:
Introduce TestStorage

ddevault committed:
Fix performance issues with song select

ddevault committed:
Update osu-framework

ddevault committed:
Fix up sqlite

peppy committed:
Add missing reference.

peppy commented on osu.Desktop.VisualTests/Program.cs:
```diff
@@ -12,7 +13,7 @@ public static class Program
         public static void Main(string[] args)
         {
-            BasicGameHost host = Host.GetSuitableHost(@"osu-visual-tests");
+            BasicGameHost host = Host.GetSuitableHost(@"osu");
```
can probably change this back now

peppy commented on osu.Game/Beatmaps/IO/BeatmapResourceStore.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace osu.Game.Beatmaps.IO
+{
+    public class BeatmapResourceStore : IResourceStore<byte[]>, IDisposable
```
is this ever used?

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+
+namespace osu.Game.GameModes.Play
+{
+    class BeatmapGroup : Container
```
Rename to `BeatmapSetPanel`?

SirCmpwn replied:
Even though it really consists of multiple discrete panels? This control has ownership over the set's panel and the difficulty panels as well.

peppy replied:
Yeah, only realised that afterwards. Let's leave this one for now.

peppy commented on osu.Game/GameModes/Play/BeatmapButton.cs:
```diff
@@ -0,0 +1,116 @@
+
+namespace osu.Game.GameModes.Play
+{
+    class BeatmapButton : Container
```
Rename to `BeatmapPanel`?

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+            }
+        }
+        
+        private void updateSelected(BeatmapInfo map)
```
also i feel like this method should go inside the property setter rather than be on its own.

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+                        if (difficultyWidth > 0.8f) difficultyWidth -= 0.025f;
+                        return new BeatmapButton(this.BeatmapSet, b)
+                        {
+                            MapSelected = beatmap => updateSelected(beatmap),
```
should be able to change this to `MapSelected = updateSelected`

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+            var buttons = difficulties.Children.ToList();
+            for (int i = 0; i < buttons.Count; i++)
+            {
+                var button = buttons[i] as BeatmapButton;
```
i'd rather this logic was just removed for now as it feels really weird.

SirCmpwn replied:
Just the logic that sets the width or the selected logic as a whole?

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+                if (collapsed == value)
+                    return;
+                collapsed = value;
+                this.ClearTransformations();
```
`this.` not needed

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+                collapsed = value;
+                this.ClearTransformations();
+                const float uncollapsedAlpha = 1;
+                Transforms.Add(new TransformAlpha(Clock)
```
`FadeTo(collapsed ? collapsedAlpha : uncollapsedAlpha, 250);`

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+                    }
+                }
+            };
+            difficulties = new FlowContainer // Deliberately not added to children
```
i'd just add this from the start and make it invisible. no real overhead from doing so.

SirCmpwn replied:
What's the best way to make it both invisible and occupy no space?

peppy replied:
Alpha = 0 is the best way at the moment (or alternatively call `Hide()`).

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+                        // TODO: Make these italic
+                        new SpriteText
+                        {
+                            Text = this.beatmapSet.Metadata.Title ?? this.beatmapSet.Metadata.TitleUnicode,
```
`this.` not needed

peppy replied:
fair call

peppy commented on osu.Game/GameModes/Play/BeatmapGroup.cs:
```diff
@@ -0,0 +1,246 @@
+                        new FlowContainer
+                        {
+                            AutoSizeAxes = Axes.Both,
+                            Children = new[]
```
I'd make this a new class `DifficultyIcon : Container`

peppy commented on osu.Game/GameModes/Play/PlaySongSelect.cs:
```diff
@@ -4,37 +4,190 @@
-                typeof(ModSelect),
-                typeof(Player)
-        };
+        private void selectBeatmapSet(BeatmapSetInfo beatmapSet)
```
private methods below ctor please.

peppy commented on osu.Game/GameModes/Play/PlaySongSelect.cs:
```diff
@@ -4,37 +4,190 @@
+        private void selectBeatmapSet(BeatmapSetInfo beatmapSet)
+        {
+            selectedBeatmapSet = beatmapSet;
+            foreach (var child in setList.Children)
```
this should be logic inside the BeatmapGroup drawable itself.

peppy commented on osu.Game/GameModes/Play/PlaySongSelect.cs:
```diff
@@ -4,37 +4,190 @@
+            selectedBeatmap = beatmap;
+        }
+
+        private Stopwatch watch = new Stopwatch();
```
remove this

peppy commented on osu.Game/GameModes/Play/PlaySongSelect.cs:
```diff
@@ -4,37 +4,190 @@
+            Scheduler.Add(() =>
+            {
+                var group = new BeatmapGroup(beatmapSet);
+                group.SetSelected += selectBeatmapSet;
```
i'd consider making these non-events and using object initialiser to bind them (just for code elegance). pretty sure we'll only be binding from one place

```
var group = new BeatmapGroup(beatmapSet)
                {
                    SetSelected = selectBeatmapSet,
                    BeatmapSelected = selectBeatmap
                };
```

ddevault committed:
Merge branch 'master' into song-select

ddevault committed:
Fix use of obsolete NUnit attribute

ddevault committed:
General cleanup per PR feedback

ddevault committed:
Use IStateful and improve collapse logic

SirCmpwn commented:
This should be good to go now.

peppy committed:
Tidy namespaces.

peppy committed:
Remove difficulty width offsets and tidy up flow of information.

peppy committed:
Change panel selection logic to avoid looping.

peppy committed:
beatmapSet can be private.

peppy committed:
Move BeatmapGroupState to outside class definition.
